src/notebook/helpers.ts

```diff
@@ -0,0 +1,42 @@
+import { NotebookDocument, workspace, WorkspaceEdit } from 'vscode'
+
+/**
+ * Use this to perform updates on all cells.
+ * We cannot update cells in parallel, this could result in data loss.
+ * E.g. assume we update execution order, while that's going on, assume we update the output (as output comes back from jupyter).
+ * At this point, VSC is still updating the execution order & we then update the output.
+ * Depending on the sequence its possible for some of the updates to get lost.
+ *
+ * Excellent example:
+ * Assume we perform the following updates without awaiting on the promise.
+ * Without awaiting, its very easy to replicate issues where the output is never displayed.
+ * - We update execution count
+ * - We update output
+ * - We update status after completion
+ */
+const pendingCellUpdates = new WeakMap<NotebookDocument, Promise<unknown>>()
```

Comment:
Hopefully by next week all of these work arounds will be a thing of the past, VSCode is updating their API (changes to how execution is handled and output is updated).

Reply:
This is one of the reasons why outputs might not appear at times,

Reply:
![julia](https://user-images.githubusercontent.com/1948812/111004352-c10c8900-833d-11eb-8d41-ba4ef5474296.gif)